microprofile/config/src/main/java/io/helidon/microprofile/config/FieldTypes.java

```diff
@@ -0,0 +1,161 @@
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Provider;
+
+/**
+ * A three tier description of a field type (main type, first
+ * generic type, second generic type).
+ */
+final class FieldTypes {
+    // such as List
+    // if this is not a generic type, all three fields will be the same value
+    private TypedField field0;
+    // such as Optional
+    private TypedField field1;
+    // such as String
+    private TypedField field2;
+
+    static FieldTypes create(Type type) {
+        FieldTypes ft = new FieldTypes();
+
+        // if the first type is a Provider or an Instance, we do not want it and start from its child
+        TypedField firstType = getTypedField(type);
+        if (Provider.class.isAssignableFrom(firstType.rawType)) {
+            ft.field0 = getTypedField(firstType);
+            firstType = ft.field0;
+        } else {
+            ft.field0 = firstType;
+        }
+
+        ft.field1 = getTypedField(ft.field0);
+
+        // now suppliers, optionals may have two levels deep
+        if (firstType.rawType == Optional.class || firstType.rawType == Supplier.class) {
+            ft.field2 = getTypedField(ft.field1);
+        } else {
+            ft.field2 = ft.field1;
+        }
+
+        return ft;
+    }
+
+    private static TypedField getTypedField(Type type) {
+        if (type instanceof Class) {
+            return new TypedField((Class<?>) type);
+        } else if (type instanceof ParameterizedType) {
+            ParameterizedType paramType = (ParameterizedType) type;
+
+            return new TypedField((Class<?>) paramType.getRawType(), paramType);
+        }
+
+        throw new UnsupportedOperationException("No idea how to handle " + type);
+    }
+
+    private static TypedField getTypedField(TypedField field) {
+        if (field.isParameterized()) {
+            ParameterizedType paramType = field.paramType;
+            Type[] typeArgs = paramType.getActualTypeArguments();
+
+            if (typeArgs.length == 1) {
+                Type typeArg = typeArgs[0];
+                return getTypedField(typeArg);
+            }
+
+            if ((typeArgs.length == 2) && (field.rawType == Map.class)) {
+                if ((typeArgs[0] == typeArgs[1]) && (typeArgs[0] == String.class)) {
+                    return new TypedField(String.class);
+                }
+            }
+
+            throw new DeploymentException("Cannot create config property for " + field.rawType + ", params: " + Arrays
+                    .toString(typeArgs));
+        }
+
+        return field;
+    }
+
+    TypedField field0() {
+        return field0;
+    }
+
+    TypedField field1() {
+        return field1;
+    }
+
+    TypedField field2() {
+        return field2;
+    }
+
+    static final class TypedField {
```

Comment:
Could this be a record too?

Reply:
This is an existing class extracted to upper level. The class was not modified other than changing visibility to work with an extracted class.

Reply:
Fixed (as this removes a lot of code)